mbzirc_ign/src/WaveVisual.cc

```diff
@@ -0,0 +1,351 @@
+  public: void OnUpdate();
+};
+
+/////////////////////////////////////////////////
+WaveVisual::WaveVisual()
+    : System(), dataPtr(std::make_unique<WaveVisualPrivate>())
+{
+}
+
+/////////////////////////////////////////////////
+WaveVisual::~WaveVisual()
+{
+}
+
+/////////////////////////////////////////////////
+void WaveVisual::Configure(const Entity &_entity,
+               const std::shared_ptr<const sdf::Element> &_sdf,
+               EntityComponentManager &_ecm,
+               EventManager &_eventMgr)
+{
+  IGN_PROFILE("WaveVisual::Configure");
+  // Ugly, but needed because the sdf::Element::GetElement is not a const
+  // function and _sdf is a const shared pointer to a const sdf::Element.
+  auto sdf = const_cast<sdf::Element *>(_sdf.get());
+
+  if (!sdf->HasElement("wavefield"))
+  {
+    ignerr << "<wavefield> parameter is missing " << std::endl;
+    return;
+  }
+  if (!sdf->HasElement("shader"))
+  {
+    ignerr << "<shader> parameter is missing " << std::endl;
+    return;
+  }
+
+  this->dataPtr->wavefield.Load(_sdf);
+
+  if (this->dataPtr->modelPath.empty())
+  {
+    auto modelEntity = topLevelModel(_entity, _ecm);
+    this->dataPtr->modelPath =
+        _ecm.ComponentData<components::SourceFilePath>(modelEntity).value();
+  }
+
+  // parse path to shaders
+  sdf::ElementPtr shaderElem = sdf->GetElement("shader");
+  if (!shaderElem->HasElement("vertex") ||
+      !shaderElem->HasElement("fragment"))
+  {
+    ignerr << "<shader> must have <vertex> and <fragment> sdf elements"
+           << std::endl;
+  }
+  else
+  {
+    sdf::ElementPtr vertexElem = shaderElem->GetElement("vertex");
+    this->dataPtr->vertexShaderUri = common::findFile(
+        asFullPath(vertexElem->Get<std::string>(), this->dataPtr->modelPath));
+    sdf::ElementPtr fragmentElem = shaderElem->GetElement("fragment");
+    this->dataPtr->fragmentShaderUri = common::findFile(
+        asFullPath(fragmentElem->Get<std::string>(), this->dataPtr->modelPath));
+  }
+
+  this->dataPtr->entity = _entity;
+  auto nameComp = _ecm.Component<components::Name>(_entity);
+  this->dataPtr->visualName = nameComp->Data();
+
+  // connect to the SceneUpdate event
+  // the callback is executed in the rendering thread so do all
+  // rendering operations in that thread
+  this->dataPtr->connection =
+      _eventMgr.Connect<ignition::gazebo::events::SceneUpdate>(
+      std::bind(&WaveVisualPrivate::OnUpdate, this->dataPtr.get()));
+}
+
+//////////////////////////////////////////////////
+void WaveVisual::PreUpdate(
+  const ignition::gazebo::UpdateInfo &_info,
+  ignition::gazebo::EntityComponentManager &)
+{
+  IGN_PROFILE("WaveVisual::PreUpdate");
+  std::lock_guard<std::mutex> lock(this->dataPtr->mutex);
+  this->dataPtr->currentSimTime = _info.simTime;
+}
+
+//////////////////////////////////////////////////
+void WaveVisualPrivate::OnUpdate()
+{
+  std::lock_guard<std::mutex> lock(this->mutex);
```

Comment:
Should we move this `lock_guard` only around the place where we read `currentSimTime`?

Reply:
moved. 6b1390777afb53df8a9011cefd6b4649eb19baf9